### PR TITLE
[lldb] Add a callback version of TCPSocket::Accept

### PR DESCRIPTION
**Pull request by labath**
The existing function already used the MainLoop class, which allows one to wait on multiple events at once. It needed to do this in order to wait for v4 and v6 connections simultaneously. However, since it was creating its own instance of MainLoop, this meant that it was impossible to multiplex these sockets with anything else.

This patch simply adds a version of this function which uses an externally provided main loop instance, which allows the caller to add any events it deems necessary. The previous function becomes a very thin wrapper over the new one.

### CODE REVIEW

**Comment by slydiman** [comment]
@labath 
> Basically, keep the socket code inside the socket class, but make it asynchronous. This way you can listen for as many sockets (and other things) as you like. It shouldn't be necessary to do the socket handling stuff externally.
> 
> I haven't looked into exactly how this would fit into the Acceptor class in lldb-server, but my preferred solutions would be:
> 
> * delete it (or at least significantly slim down it's functionality -- it's a very thin wrapper over the Socket class, and probably only exists due to historic reasons)
> * make it callback-based as well
>
> I don't like the idea of making Acceptor responsible for multiple ports for the same reason I did not want to do that for TCPSocket -- it's complicated and inflexible.

I don't see any profit of this change. Using an extrernal MainLoop we can break the blocking Accept() from an other thread. But we can break it making a dummy connection too. 

The main reason to change TCPSocket was to be able to wait for multiple connections from different ports in the same blocking Accept() in the single thread.
Currently #104238 contains the best implementation of Acceptor.cpp w/o changing TCPSocket.

sock_cb parameter is useles here because it will be always the same. 
TCPSocket::m_listen_sockets is still here and it is still inaccesible externally. The method Listen() which initializes m_listen_sockets is unchanged. 
A callback may be useful (in theory) in Listen() to add a custom listen sockets to m_listen_sockets. Probably it is better to just initialize an own listen_sockets and just provide it to TCPSocket. But where this code will be placed? Acceptor.cpp, lldb-platform.cpp?

If you want to delete Acceptor, the best approach is the following:

Use Listen() from #104797 but w/o changing the connection string:
Status TCPSocket::Listen(llvm::StringRef name, int backlog, std::set<uint16_t> * extra_ports = nullptr)

The accept MainLoop must be a member of TCPSocket and it is enough to just add 
TCPSocket::BreakAccept() { m_accept_loop.AddPendingCallback([](MainLoopBase &loop) { loop.RequestTermination(); }); } 

> I haven't looked into exactly how this would fit into the Acceptor class in lldb-server

This is the problem. I don't see how this PR will help to implement accepting for multiple connections from 2 ports. It only complicates TCPSocket.

I'm trying to offer the `completed` solution which has been tested.

BTW, please approve NFC #106640.

**Comment by slydiman** [comment]
> I don't like the idea of making Acceptor responsible for multiple ports for the same reason I did not want to do that for TCPSocket -- it's complicated and inflexible.

Note Acceptor is used ONLY in lldb-platform.cpp
We can use the Accept() for the second port in the second thread. Then it is enough to add TCPSocket::BreakAccept() or just use a dummy connection and don't change TCPSocket and Accptor at all. 
But the common Accept() for both connections in the single thread looks much better.

**Comment by labath** [comment]
> @labath
> 
> > Basically, keep the socket code inside the socket class, but make it asynchronous. This way you can listen for as many sockets (and other things) as you like. It shouldn't be necessary to do the socket handling stuff externally.
> > I haven't looked into exactly how this would fit into the Acceptor class in lldb-server, but my preferred solutions would be:
> > 
> > * delete it (or at least significantly slim down it's functionality -- it's a very thin wrapper over the Socket class, and probably only exists due to historic reasons)
> > * make it callback-based as well
> > 
> > I don't like the idea of making Acceptor responsible for multiple ports for the same reason I did not want to do that for TCPSocket -- it's complicated and inflexible.
> 
> I don't see any profit of this change. Using an extrernal MainLoop we can break the blocking Accept() from an other thread. But we can break it making a dummy connection too.

This wasn't my main motivation, but now that you mention it, I actually think this may be reason enough to do this. Using a dummy connection to break out of a blocking accept is.. clever, but I wouldn't consider it a good practice. I also think we have other places in lldb that could use the ability to interrupt a blocking operation, and I wouldn't want to start adding dummy connections there as well. (And I'm pretty sure I can come up with a (contrived) scenario where that dummy connect will actually fail.)

> 
> The main reason to change TCPSocket was to be able to wait for multiple connections from different ports in the same blocking Accept() in the single thread. 

In a single thread -- yes. In the same blocking accept -- not really, at least not for me.

> Currently #104238 contains the best implementation of Acceptor.cpp w/o changing TCPSocket.

I guess we'll have to agree to disagree on that.

> 
> sock_cb parameter is useles here because it will be always the same.

I'll have to disagree with that. The reason we have two ports is because we want to use them for different things. One of them will be launching a platform instance, the other a gdbserver instance. So, even though the logic will be roughly similar (both are spawning a new process) each will be doing something slightly different. With two sockets, you can pass one callback to the gdbserver socket and a different one to the platform socket.

But even if the callback was the same every time, I'd still think this is a good separation between "how to accept a connection" and "what to do with the accepted connection".

> TCPSocket::m_listen_sockets is still here and it is still inaccesible externally. The method Listen() which initializes m_listen_sockets is unchanged.

Yes, and there's no reason to change that. (I'd definitely like to change that, but that's not necessary here)

> A callback may be useful (in theory) in Listen() to add a custom listen sockets to m_listen_sockets. Probably it is better to just initialize an own listen_sockets and just provide it to TCPSocket. But where this code will be placed? Acceptor.cpp, lldb-platform.cpp?

Now I'm just thinking we haven't understood each other about how this could be used. There's no reason to pass additional sockets, because there's no reason to have a unified socket list (well.. there is, but it's already present in the main loop). To listen on two (or however many) ports, you just create however many TCPSocket instances you want, and then do a mainloop.Run(). And yeah, I'd ideally place the code in lldb-platform.cpp (that's the proper place for a main loop).

The code would be something like:
```
MainLoop loop;
platform_socket->Accept(loop, spawn_platform_cb);
server_socket->Accept(loop, spawn_gdbserver_cb);
loop.Run();
```
> 
> If you want to delete Acceptor, the best approach is the following:
I do, but it's not my primary goal. Definitely not at the cost of putting a MainLoop member inside TCPSocket.

If it turns out to be easier, I'd be fine with turning `platform_socket` and `server_socket` instances from the above snippet into `Acceptor`s.

File: lldb/unittests/Host/SocketTest.cpp [hunk 1]
```diff
@@ -95,6 +96,40 @@ TEST_P(SocketTest, TCPListen0ConnectAccept) {
                             &socket_b_up);
 }
 
+TEST_P(SocketTest, TCPMainLoopAccept) {
+  const bool child_processes_inherit = false;
+  auto listen_socket_up =
+      std::make_unique<TCPSocket>(true, child_processes_inherit);
+  Status error = listen_socket_up->Listen(
+      llvm::formatv("[{0}]:0", GetParam().localhost_ip).str(), 5);
```
Comment (slydiman) [inline lldb/unittests/Host/SocketTest.cpp:104]: It seems extra brackets cause the failure in case of IPv6.

**Review by slydiman — APPROVED** [review]
Ok, let's move on this way.
But fix the test please.

**Review by slydiman — CHANGES_REQUESTED** [review]
BTW, I have figured out a side effect. MainLoopWindows uses events created by WSACreateEvent(). WSACreateEvent() creates the `manual-reset` event. But WSAResetEvent() is used only for m_trigger_event. All other events (READ, ACCEPT, CLOSE) are never reset. I have no idea how it worked before. But I got thousands of error messages using this patch with lldb-server after first connection:
```
AcceptSocket(): A non-blocking socket operation could not be completed immediately
```
I have added `WSAResetEvent(KV.second.event);` before `ProcessReadObject(KV.first);` in MainLoopWindows.cpp to fix it. It seems this issue must be fixed before or with this patch, otherwise it will break the Windows build.

**Review by slydiman — APPROVED** [review]
LGTM, thanks

**Pull request merged**